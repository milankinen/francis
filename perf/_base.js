const memwatch = require("@airbnb/node-memwatch")
const Benchmark = require("benchmark")
const bytes = require("pretty-bytes")
const { format } = require("util")
const { Francis, Bacon, Kefir } = require("./_libs")

const MEMORY = !!process.env.SHOW_MEMORY_USAGE
let heapStats = []
let gcTime = 0

const withFrancis = f => (...args) => f(Francis, ...args)
const withBacon = f => (...args) => f(Bacon, ...args)
const withKefir = f => (...args) => f(Kefir, ...args)

const resetHeapStats = () => {
  gcTime = 0
  heapStats = []
}

const getAvgMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => a + s, 0) / stats.length)) : "unknown"
const getMaxMemoryUsage = stats =>
  stats.length ? bytes(Math.floor(stats.reduce((a, s) => Math.max(a, s), 0))) : "unknown"

if (MEMORY) {
  memwatch.on("stats", stats => {
    heapStats.push(stats.used_heap_size)
    gcTime += stats.gc_time / 1000000.0
  })
}

const testCases = (label, fn, combinations = []) => {
  const accepted = (testFn, args) => !testFn.accept || testFn.accept(...args)
  const cases = (name, testFn) =>
    combinations
      .map(
        args =>
          accepted(testFn, args) && {
            description:
              `${name} ${format(label + "|||", ...args).split("|||")[0]}` +
              (testFn.note ? ` (${testFn.note})` : ""),
            defer: true,
            fn: d => {
              const done = () => d.resolve()
              testFn(...args, done)
            },
          },
      )
      .filter(x => x)
  return [
    ...(process.env.NO_FRANCIS ? [] : cases("francis", withFrancis(fn))),
    ...(process.env.NO_BACON ? [] : cases("bacon  ", withBacon(fn))),
    ...(process.env.NO_KEFIR || !fn.kefir ? [] : cases("kefir  ", withKefir(fn.kefir))),
  ]
}

const runTests = (suiteName, testCases) => {
  return new Promise(resolve => {
    console.log(`Running suite ${suiteName}...`)
    console.log("")
    const maxLen = testCases.reduce((m, t) => Math.max(t.description.length, m), 0)
    resetHeapStats()
    testCases
      .reduce(
        (suite, test) => suite.add(test.description.padEnd(maxLen + 1), test),
        new Benchmark.Suite(),
      )
      .on("complete", function() {
        if (MEMORY) {
          const stats = [...heapStats]
          console.log("Memory statistics:")
          console.log(" > Avg mem:", getAvgMemoryUsage(stats))
          console.log(" > Max mem:", getMaxMemoryUsage(stats))
          console.log(" > GC time:", gcTime.toFixed(2), "ms")
        } else {
          this.forEach(test => console.log(test.toString()))
        }
        console.log("")
        resolve()
      })
      .run({ async: true })
  })
}

module.exports = {
  testCases,
  runTests,
}
