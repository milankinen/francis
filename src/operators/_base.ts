import {
  NOOP_SUBSCRIBER,
  NOOP_SUBSCRIPTION,
  sendEndSafely,
  sendErrorSafely,
  sendEventSafely,
  sendInitialSafely,
  sendNoInitialSafely,
  Source,
  Subscriber,
  Subscription,
} from "../_core"
import { MAX_ORDER } from "../_priority"
import { Transaction } from "../_tx"
import { disableNoUnusedWarning } from "../_util"
import { Scheduler, Task } from "../scheduler/index"

export enum EventType {
  INITIAL = 1,
  NO_INITIAL,
  EVENT,
  ERROR,
  END,
}

/**
 * This function is meant for actual sources (e.g. once, fromArray etc..) to
 * enable multicasting (thus keeping the source's codebase clean)
 *
 * @param source Unicasted "bare" source
 */
export function identity<T>(source: Source<T>): Operator<T, T> {
  return new Identity(source)
}

export abstract class MulticastImplementation<T> implements Subscriber<T> {
  public readonly next: Subscriber<T> = NOOP_SUBSCRIBER

  /**
   * This hook is called every time when operator is activated (gets its first
   * subscriber). If this function returns false, operator aborts the current
   * subscription process. The impelementation may also queue tasks to the given
   * scheduler before returning.
   *
   * @param scheduler
   * @param subscriber
   */
  public beforeActivation(scheduler: Scheduler, subscriber: Subscriber<T>): boolean {
    return true
  }

  /**
   * This hook is called every time when operator is multicasted (gets a new subscriber
   * after initial activation). If this function returns false, operator aborts the current
   * subscription process. The impelementation may also queue tasks to the given
   * scheduler before returning.
   *
   * @param scheduler
   * @param subscriber
   */
  public beforeMulticast(scheduler: Scheduler, subscriber: Subscriber<T>): boolean {
    return true
  }

  public isActive(): boolean {
    return true
  }

  public initial(tx: Transaction, val: T): void {
    this.next.initial(tx, val)
  }

  public noinitial(tx: Transaction): void {
    this.next.noinitial(tx)
  }

  public event(tx: Transaction, val: T): void {
    this.next.event(tx, val)
  }

  public error(tx: Transaction, err: Error): void {
    this.next.error(tx, err)
  }

  public end(tx: Transaction): void {
    this.next.end(tx)
  }
}

const NOOP_MCAST = new class NoopImpl extends MulticastImplementation<any> {}()

export abstract class Operator<A, B>
  implements Subscriber<A>, Source<B>, AbortSubscriptionListener<B> {
  public readonly weight: number

  protected readonly order: number
  protected readonly source: Source<A>
  protected readonly next: MulticastImplementation<B> = NOOP_MCAST

  private __subs: Subscription = NOOP_SUBSCRIPTION
  private __active: boolean = false

  constructor(origin: Source<A>) {
    false && disableNoUnusedWarning(this.__handleDispose, this.__handeReorder)
    this.source = origin
    this.order = MAX_ORDER + 1
    this.weight = origin.weight
  }

  /**
   * Called only internally from Property and EventStream constructors
   * @param impl Multicast implementation to use for this operator
   */
  public setMulticastImplementation(impl: MulticastImplementation<B>): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.next as MulticastImplementation<B>) = impl
  }

  /**
   * Called only internally when new EventStream or Property subscribes to this operator
   * during their activation.
   * @param scheduler
   * @param subscriber
   * @param order
   */
  public subscribe(scheduler: Scheduler, subscriber: Subscriber<B>, order: number): Subscription {
    const next = this.next.next
    return next === NOOP_SUBSCRIBER
      ? this.__handleActivation(scheduler, subscriber, order)
      : this.__handleMulticast(scheduler, subscriber, order)
  }

  public abstract initial(tx: Transaction, val: A): void

  public abstract event(tx: Transaction, val: A): void

  public noinitial(tx: Transaction): void {
    this.next.noinitial(tx)
  }

  public error(tx: Transaction, err: Error): void {
    this.next.error(tx, err)
  }

  public end(tx: Transaction): void {
    this.next.end(tx)
  }

  public isActive(): boolean {
    return this.__active
  }

  public handleAbort(subscriber: Subscriber<B>): void {}

  protected activateFirst(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    // TODO
    this.__active = true
    this.__subs = this.source.subscribe(scheduler, this, order)
    return new MulticastSubscription(subscriber, this as any)
  }

  protected activateLate(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return new MulticastSubscription(subscriber, this as any)
  }

  protected abort(scheduler: Scheduler, subscriber: Subscriber<B>, order: number): Subscription {
    scheduler.scheduleAbortSubscription(new AbortSubscriptionTask(this, subscriber))
    return this.activateLate(scheduler, subscriber, order)
  }

  protected propagateReorder(order: number): void {
    this.__subs.reorder(order)
  }

  protected dispose(): void {
    this.__active = false
    this.__subs.dispose()
    this.__subs = NOOP_SUBSCRIPTION
  }

  protected handleActivation(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return this.activateFirst(scheduler, subscriber, order)
  }

  protected handleMulticast(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    return this.activateLate(scheduler, subscriber, order)
  }

  protected handleReorder(order: number): void {
    this.propagateReorder(order)
  }

  protected handleDispose(): void {
    this.dispose()
  }

  private __updateNext(next: Subscriber<B>): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.next.next as Subscriber<B>) = next
  }

  private __updateOrder(order: number): void {
    // tslint:disable-next-line:semicolon whitespace
    ;(this.order as number) = order
  }

  private __handleActivation(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    if (this.next.beforeActivation(scheduler, subscriber)) {
      this.__updateNext(subscriber)
      this.__updateOrder(order)
      return this.handleActivation(scheduler, subscriber, order)
    } else {
      return NOOP_SUBSCRIPTION
    }
  }

  private __handleMulticast(
    scheduler: Scheduler,
    subscriber: Subscriber<B>,
    order: number,
  ): Subscription {
    if (this.next.beforeMulticast(scheduler, subscriber)) {
      const next = this.next.next
      if (next instanceof MulticastDelegatee) {
        next.add(subscriber, order)
      } else {
        this.__updateNext(
          new MulticastDelegatee({ s: next, o: this.order }, { s: subscriber, o: order }),
        )
      }
      if (order < this.order) {
        this.__updateOrder(order)
        this.handleReorder(order)
      }
      return this.handleMulticast(scheduler, subscriber, order)
    } else {
      return NOOP_SUBSCRIPTION
    }
  }

  private __handleDispose(subscriber: Subscriber<B>): void {
    const next = this.next.next
    if (next instanceof MulticastDelegatee) {
      const { o: order, s: updated } = next.remove(subscriber)
      this.__updateNext(updated)
      if (order !== this.order) {
        this.__updateOrder(order)
        this.handleReorder(order)
        this.__subs.reorder(order)
      }
    } else {
      this.__updateNext(NOOP_SUBSCRIBER)
      this.__updateOrder(MAX_ORDER + 1)
      this.handleDispose()
    }
  }
  private __handeReorder(subscriber: Subscriber<B>, order: number): void {
    const next = this.next.next
    next instanceof MulticastDelegatee && next.updateOrder(subscriber, order)
    if (order < this.order) {
      this.__updateOrder(order)
      this.handleReorder(order)
    }
  }
}

export interface AbortSubscriptionListener<T> {
  handleAbort(subsciber: Subscriber<T>): void
}

export class AbortSubscriptionTask<T> implements Task {
  constructor(private li: AbortSubscriptionListener<T>, private subscriber: Subscriber<T>) {}

  public run(): void {
    this.li.handleAbort(this.subscriber)
  }
}

const NIL: OrderedSubscriberNode<any> = {
  a: false,
  n: null as any,
  o: MAX_ORDER + 1,
  s: NOOP_SUBSCRIBER,
}

class MulticastDelegatee<T> implements Subscriber<T> {
  private n: number
  private head: OrderedSubscriberNode<T>
  private tail: OrderedSubscriberNode<T>

  constructor(a: OrderedSubscriber<T>, b: OrderedSubscriber<T>) {
    this.head = this.tail = NIL
    this.head = { ...a, a: true, n: this.head }
    this.head = { ...b, a: true, n: this.head }
    this.n = 2
  }

  public add(subscriber: Subscriber<T>, order: number): void {
    this.head = { o: order, s: subscriber, a: true, n: this.head }
    ++this.n
  }

  public remove(subscriber: Subscriber<T>): OrderedSubscriber<T> {
    let prev = NIL
    let next = this.head
    let order = MAX_ORDER + 1
    while (next !== NIL) {
      if (next.s === subscriber) {
        next.a = false
        prev !== NIL ? (prev.n = next.n) : (this.head = next.n)
        --this.n
      } else if (next.o < order) {
        order = next.o
      }
      prev = next
      next = next.n
    }
    return {
      o: order,
      s: this.n > 1 ? this : this.head.s,
    }
  }

  public updateOrder(subscriber: Subscriber<T>, order: number): void {
    let next = this.head
    while (next !== NIL) {
      if (next.s === subscriber) {
        next.o = order
        return
      } else {
        next = next.n
      }
    }
  }

  public isActive(): boolean {
    return true
  }

  public initial(tx: Transaction, val: T): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendInitialSafely(tx, next.s, val)
      next = next.n
    }
  }

  public noinitial(tx: Transaction): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendNoInitialSafely(tx, next.s)
      next = next.n
    }
  }

  public event(tx: Transaction, val: T): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendEventSafely(tx, next.s, val)
      next = next.n
    }
  }

  public error(tx: Transaction, err: Error): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendErrorSafely(tx, next.s, err)
      next = next.n
    }
  }

  public end(tx: Transaction): void {
    let next = this.head
    while (next !== NIL) {
      next.a && sendEndSafely(tx, next.s)
      next = next.n
    }
  }
}

interface OrderedSubscriberNode<T> extends OrderedSubscriber<T> {
  n: OrderedSubscriberNode<T>
  a: boolean
}

class MulticastSubscription<T> implements Subscription {
  private s: Subscriber<T>
  private o: MulticastSubscriptionOwner<T>

  // tslint:disable-next-line:ban-types
  constructor(subscriber: Subscriber<T>, owner: MulticastSubscriptionOwner<T>) {
    this.s = subscriber
    this.o = owner
  }

  public dispose(): void {
    this.o.__handleDispose(this.s)
  }

  public reorder(order: number): void {
    this.o.__handeReorder(this.s, order)
  }
}

interface MulticastSubscriptionOwner<T> {
  __handleDispose(subscriber: Subscriber<T>): void
  __handeReorder(subscriber: Subscriber<T>, order: number): void
}

interface OrderedSubscriber<T> {
  s: Subscriber<T>
  o: number
}

class Identity<T> extends Operator<T, T> {
  public initial(tx: Transaction, val: T): void {
    this.next.initial(tx, val)
  }
  public event(tx: Transaction, val: T): void {
    this.next.event(tx, val)
  }
}
