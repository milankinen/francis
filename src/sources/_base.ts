import { __DEVELOPER__, logAndThrow } from "../_assert"
import {
  NOOP_SUBSCRIBER,
  sendRootEnd,
  sendRootError,
  sendRootNext,
  Source,
  Subscriber,
  Subscription,
} from "../_core"
import { AnyEvent } from "../_interfaces"
import { isEnd, isError, isEvent, isNext } from "../Event"
import { scheduleActivationTask, Task } from "../scheduler/index"

export abstract class Root<T> implements Source<T> {
  public readonly weight: number = 1
  public ended: boolean = false
  constructor(public readonly sync: boolean) {}

  public subscribe(subscriber: Subscriber<T>, order: number): Subscription {
    return this.create(subscriber)
  }

  protected abstract create(subscriber: Subscriber<T>): Activation<T, any>
}

export abstract class Activation<T, R extends Root<T>> implements Task, Subscription {
  // the real subscriber
  protected readonly real: Subscriber<T>
  protected readonly owner: R
  protected active: boolean = true
  protected subscriber: Subscriber<T>

  constructor(owner: R, subscriber: Subscriber<T>) {
    this.real = this.subscriber = subscriber
    this.owner = owner
  }

  public run(): void {
    if (this.active) {
      if (this.owner.ended) {
        sendRootEnd(this.subscriber)
      } else {
        if (this.subscriber.begin()) {
          this.start()
        }
      }
    }
  }

  public activate(): void {
    if (this.owner.sync) {
      this.run()
    } else {
      scheduleActivationTask(this)
    }
  }

  public dispose(): void {
    this.active = false
    this.subscriber = NOOP_SUBSCRIBER
    this.stop()
  }

  public reorder(order: number): void {
    // no-op
  }

  protected abstract start(): void

  protected abstract stop(): void

  protected send(event: T | AnyEvent<T>): void {
    if (isEvent(event)) {
      if (isNext(event)) {
        this.sendNext(event.value)
      } else if (isError(event)) {
        this.sendError(event.error)
      } else if (isEnd(event)) {
        this.sendEnd()
      } else {
        if (__DEVELOPER__) {
          logAndThrow("**BUG** event type not known: " + event)
        }
      }
    } else {
      sendRootNext(this.subscriber, event)
    }
  }

  protected sendNext(val: T): void {
    sendRootNext(this.subscriber, val)
  }

  protected sendError(err: Error): void {
    sendRootError(this.subscriber, err)
  }

  protected sendEnd(): void {
    this.owner.ended = true
    sendRootEnd(this.subscriber)
  }
}
