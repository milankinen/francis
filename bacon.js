export * from "./dist/francis.bacon"
