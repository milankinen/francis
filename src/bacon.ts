import { assert } from "./_assert"
import { Accum, Dispose, Handler, Predicate, Projection, ValueHandler } from "./_interfaces"
import {
  argsToObservables,
  argsToObservablesAndFunction,
  toFunction,
  toFunctionsPropAsIs,
} from "./_interrop"
import { EventStream } from "./EventStream"
import { Observable } from "./Observable"
import * as Concat from "./operators/concat"
import * as Filter from "./operators/filter"
import * as First from "./operators/first"
import * as FlatMap from "./operators/flatMap"
import * as Last from "./operators/last"
import * as Log from "./operators/log"
import * as Logic from "./operators/logic"
import * as Map from "./operators/map"
import * as Merge from "./operators/merge"
import * as OnValue from "./operators/onValue"
import * as Sample from "./operators/sample"
import * as Scan from "./operators/scan"
import * as StartWith from "./operators/startWith"
import * as Subscribe from "./operators/subscribe"
import * as Take from "./operators/take"
import * as ToEventStream from "./operators/toEventStream"
import * as ToProperty from "./operators/toProperty"
import * as Zip from "./operators/zip"
import { Property } from "./Property"

declare module "./Observable" {
  interface Observable<A> {
    subscribe(handler: Handler<A>): Dispose
    onValue(f: ValueHandler<A>): Dispose
    log(label?: string): Dispose
    map<B>(project: Projection<A, B>): Observable<B>
    filter(predicate: Predicate<A> | Property<any>): Observable<A>
    take(n: number): Observable<A>
    first(): Observable<A>
    last(): Observable<A>
    flatMapLatest<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMapFirst<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMap<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMapConcat<B>(project: Projection<A, B | Observable<B>>): Observable<B>
    flatMapWithConcurrencyLimit<B>(
      limit: number,
      project: Projection<A, B | Observable<B>>,
    ): Observable<B>
    startWith(value: A): Observable<A>
    scan<B>(seed: B, f: Accum<B, A>): Property<B>
    zip<B, C>(other: Observable<B>, f: (a: A, b: B) => C): EventStream<C>
    merge(other: Observable<A>): EventStream<A>
    concat(other: Observable<A>): EventStream<A>
  }
}

declare module "./EventStream" {
  interface EventStream<A> {
    toProperty(initialValue?: A): Property<A>
    map<B>(project: Projection<A, B>): EventStream<B>
    filter(predicate: Predicate<A> | Property<any>): EventStream<A>
    take(n: number): EventStream<A>
    first(): EventStream<A>
    last(): EventStream<A>
    flatMapLatest<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMapFirst<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMap<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMapConcat<B>(project: Projection<A, B | Observable<B>>): EventStream<B>
    flatMapWithConcurrencyLimit<B>(
      limit: number,
      project: Projection<A, B | Observable<B>>,
    ): EventStream<B>
    startWith(value: A): EventStream<A>
  }
}

declare module "./Property" {
  interface Property<A> {
    toEventStream(): EventStream<A>
    map<B>(project: Projection<A, B>): Property<B>
    filter(predicate: Predicate<A> | Property<any>): Property<A>
    take(n: number): Property<A>
    first(): Property<A>
    last(): Property<A>
    flatMapLatest<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMapFirst<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMap<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMapConcat<B>(project: Projection<A, B | Observable<B>>): Property<B>
    flatMapWithConcurrencyLimit<B>(
      limit: number,
      project: Projection<A, B | Observable<B>>,
    ): Property<B>
    sampledBy<B>(sampler: EventStream<B>): EventStream<A>
    sampledBy<B>(sampler: Property<B>): Property<A>
    sampledBy<B, C>(sampler: EventStream<B>, f: SampleFn<A, B, C>): EventStream<C>
    sampledBy<B, C>(sampler: Property<B>, f: SampleFn<A, B, C>): Property<C>
    startWith(value: A): Property<A>
    and<B>(other: Property<B>): Property<Logic.AndResult<A, B>>
    or<B>(other: Property<B>): Property<Logic.OrResult<A, B>>
    not<B>(): Property<boolean>
  }
}

export type SampleFn<V, S, R> = (value: V, sample: S) => R

/* tslint:disable:no-string-literal */

// Observable operators (common for both EventStream and Property)

Observable.prototype["subscribe"] = function<A>(handler: Handler<A>): Dispose {
  return Subscribe.subscribe(handler, this)
}

Observable.prototype["onValue"] = function<A>(f: ValueHandler<A>): Dispose {
  return OnValue.onValue(f, this)
}

Observable.prototype["log"] = function(label?: string): Dispose {
  return Log.log(label, this)
}

Observable.prototype["map"] = function<A, B>(
  project: Projection<A, B> | Property<B>,
  ...rest: any[]
): Observable<B> {
  return Map.map(toFunctionsPropAsIs(project, rest), this)
}

Observable.prototype["filter"] = function<A>(
  predicate: Predicate<A> | Property<any>,
  ...rest: any[]
): Observable<A> {
  return Filter.filter(toFunctionsPropAsIs(predicate, rest), this)
}

Observable.prototype.take = function<A>(n: number): Observable<A> {
  return Take.take(n, this)
}

Observable.prototype["first"] = function<A>(): Observable<A> {
  return First.first(this)
}

Observable.prototype["last"] = function<A>(): Observable<A> {
  return Last.last(this)
}

Observable.prototype["flatMapLatest"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapLatest(toFunction(project, rest), this)
}

Observable.prototype["flatMapFirst"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapFirst(toFunction(project, rest), this)
}

Observable.prototype["flatMap"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMap(toFunction(project, rest), this)
}

Observable.prototype["flatMapConcat"] = function<A, B>(
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapConcat(toFunction(project, rest), this)
}

Observable.prototype["flatMapWithConcurrencyLimit"] = function<A, B>(
  limit: number,
  project: Projection<A, B | Observable<B>>,
  ...rest: any[]
): Observable<B> {
  return FlatMap.flatMapWithConcurrencyLimit(limit, toFunction(project, rest), this)
}

Observable.prototype["startWith"] = function<A>(value: A): Observable<A> {
  return StartWith.startWith(value, this)
}

Observable.prototype["scan"] = function<A, B>(
  seed: B,
  f: Accum<B, A>,
  ...rest: any[]
): Property<B> {
  return Scan.scan(seed, toFunction(f, rest), this)
}

Observable.prototype["zip"] = function<A, B, C>(
  other: Observable<B>,
  f: (a: A, b: B) => C,
  ...rest: any[]
): EventStream<C> {
  return Zip.zipWith(toFunction(f, rest), [this, other])
}

Observable.prototype["merge"] = function<A>(other: Observable<A>): EventStream<A> {
  return Merge.mergeAll([this, other])
}

Observable.prototype["concat"] = function<A>(other: Observable<A>): EventStream<A> {
  return Concat.concatAll([this, other])
}

// EventStream specific operators

EventStream.prototype["toProperty"] = function<A>(initialValue?: A): Property<A> {
  return arguments.length === 0
    ? ToProperty.toProperty(this)
    : StartWith._startWithP(initialValue, ToProperty.toProperty(this))
}

const esProto = EventStream.prototype as any
esProto["toEventStream"] = function<A>(): EventStream<A> {
  return this
}

// Property specific operators

Property.prototype["sampledBy"] = function<A, B, C>(
  sampler: Observable<B>,
  f?: SampleFn<A, B, C>,
  ...rest: any[]
): any {
  const fn = toFunction(f === undefined ? (v: A, s: B) => v as any : f, rest)
  return Sample.sampleWith(sampler, fn, this)
}

Property.prototype["and"] = function<A, B>(other: Property<B>): Property<Logic.AndResult<A, B>> {
  return Logic.and(this, other)
}

Property.prototype["or"] = function<A, B>(other: Property<B>): Property<Logic.OrResult<A, B>> {
  return Logic.or(this, other)
}

Property.prototype["not"] = function<A>(): Property<boolean> {
  return Logic.not(this)
}

const pProto = Property.prototype as any
pProto["toProperty"] = function<A>(initialValue?: A): Property<A> {
  assert(arguments.length === 0, "No arguments supported")
  return this
}

Property.prototype["toEventStream"] = function<A>(): EventStream<A> {
  return ToEventStream.toEventStream(this)
}

// static operators

export { End, Error, Next } from "./Event"
export { EventStream } from "./EventStream"
export { Observable } from "./Observable"
export { combineAsArray, combineTemplate } from "./operators/combine"
export { when } from "./operators/when"
export { zipAsArray } from "./operators/zip"
export { Property } from "./Property"
export { fromArray } from "./sources/fromArray"
export { fromBinder } from "./sources/fromBinder"
export { fromPoll } from "./sources/fromPoll"
export { later } from "./sources/later"
export { never } from "./sources/never"
export { sequentially } from "./sources/sequentially"
export { constant, once } from "./sources/single"
// classes and interfaces
export * from "./_interfaces"

export function zipWith<A, T>(f: (a: A) => T, streams: [Observable<A>]): EventStream<T>
export function zipWith<A, B, T>(
  f: (a: A, b: B) => T,
  streams: [Observable<A>, Observable<B>],
): EventStream<T>
export function zipWith<A, B, C, T>(
  f: (a: A, b: B, c: C) => T,
  streams: [Observable<A>, Observable<B>, Observable<C>],
): EventStream<T>
export function zipWith<A, B, C, D, T>(
  f: (a: A, b: B, c: C, d: D) => T,
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>],
): EventStream<T>
export function zipWith<A, B, C, D, E, T>(
  f: (a: A, b: B, c: C, d: D, e: E) => T,
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>, Observable<E>],
): EventStream<T>
export function zipWith<A, B, C, D, E, F, T>(
  f: (a: A, b: B, c: C, d: D, e: E, f: F) => T,
  streams: [
    Observable<A>,
    Observable<B>,
    Observable<C>,
    Observable<D>,
    Observable<E>,
    Observable<F>
  ],
): EventStream<T>
export function zipWith<A, T>(f: (...args: A[]) => T, streams: A[]): EventStream<T>
export function zipWith<A, T>(f: (streams: [Observable<A>], a: A) => T): EventStream<T>
export function zipWith<A, B, T>(
  streams: [Observable<A>, Observable<B>],
  f: (a: A, b: B) => T,
): EventStream<T>
export function zipWith<A, B, C, T>(
  streams: [Observable<A>, Observable<B>, Observable<C>],
  f: (a: A, b: B, c: C) => T,
): EventStream<T>
export function zipWith<A, B, C, D, T>(
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>],
  f: (a: A, b: B, c: C, d: D) => T,
): EventStream<T>
export function zipWith<A, B, C, D, E, T>(
  streams: [Observable<A>, Observable<B>, Observable<C>, Observable<D>, Observable<E>],
  f: (a: A, b: B, c: C, d: D, e: E) => T,
): EventStream<T>
export function zipWith<A, B, C, D, E, F, T>(
  streams: [
    Observable<A>,
    Observable<B>,
    Observable<C>,
    Observable<D>,
    Observable<E>,
    Observable<F>
  ],
  f: (a: A, b: B, c: C, d: D, e: E, f: F) => T,
): EventStream<T>
export function zipWith<A, T>(streams: A[], f: (...args: A[]) => T): EventStream<T>
export function zipWith(...args: any[]): EventStream<any> {
  const [observables, f] = argsToObservablesAndFunction(args)
  return Zip.zipWith(f, observables as any)
}

export function concatAll<T>(...observables: Array<Observable<T>>): EventStream<T>
export function concatAll<T>(observables: Array<Observable<T>>): EventStream<T>
export function concatAll<T>(
  observables: Array<Observable<T>> | Observable<T>,
  ...rest: Array<Observable<T>>
): EventStream<T> {
  return Concat.concatAll(argsToObservables([observables, ...rest]) as Array<EventStream<T>>)
}

export function mergeAll<T>(...observables: Array<Observable<T>>): EventStream<T>
export function mergeAll<T>(observables: Array<Observable<T>>): EventStream<T>
export function mergeAll<T>(
  observables: Array<Observable<T>> | Observable<T>,
  ...rest: Array<Observable<T>>
): EventStream<T> {
  return Merge.mergeAll(argsToObservables([observables, ...rest]))
}
