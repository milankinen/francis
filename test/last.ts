import * as F from "../bacon"
import { run, Sync } from "./_base"

describe("EventStream.last", () => {
  it("results in EventStream", () => {
    expect(F.once(1).last()).toBeInstanceOf(F.EventStream)
  })

  it("takes last event and ends", () => {
    const recording = run(record =>
      F.fromArray([1, 2, 3, 4])
        .last()
        .subscribe(record),
    )
    expect(recording).toMatchSnapshot()
  })
})

describe("Property.last", () => {
  it("results in Property", () => {
    expect(F.constant(1).last()).toBeInstanceOf(F.Property)
  })

  it("takes last event and ends", () => {
    const recording = run(record =>
      F.fromArray([1, 2, 3, 4])
        .toProperty(0)
        .last()
        .subscribe(record),
    )
    expect(recording).toMatchSnapshot()
  })

  it("works with F.constant", () => {
    const recording = run(record => {
      F.constant("tsers")
        .last()
        .subscribe(record)
      record(Sync)
    })
    expect(recording).toMatchSnapshot()
  })
})
