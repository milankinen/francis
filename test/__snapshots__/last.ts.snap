// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventStream.last takes last event and ends 1`] = `
Array [
  Next {
    "value": 4,
  },
  End {},
]
`;

exports[`Property.last takes last event and ends 1`] = `
Array [
  Next {
    "value": 4,
  },
  End {},
]
`;

exports[`Property.last works with F.constant 1`] = `
Array [
  Next {
    "value": "tsers",
  },
  End {},
  Sync {},
]
`;
